flow/project.py

```diff
@@ -236,6 +251,119 @@ def make_bundles(operations, size=None):
     return _make_bundles(operations, size)
 
 
+class aggregate(object):
+    """Decorator for operation functions that are aggregate operations.
+
+    .. code-block:: python
+        example_aggregate = aggregate()
+
+        @example_aggregate
+        @FlowProject.operation
+        def foo(jobs):
+            return len(jobs)
+
+    :param groupper:
+        Information on how to aggregate jobs.
+    :type groupper:
+        callable
+    :param sort:
+        Sort jobs by a statepoint parameter.
+    :type sort:
+        str
+    :param reverse:
+        States whether sorting is to be done in ascending or descending order.
+    :type reverse:
+        bool
+    """
+
+    def __init__(self, grouper=None, sort=None, reverse=False):
+        if grouper is None:
+            def grouper(jobs):
+                yield jobs
+
+        def key_sort(job):
+            try:
+                return job.sp[sort]
+            except KeyError:
+                raise KeyError("The key '{}' was not found in statepoint "
+                               "parameters of the job {}.".format(sort, job))
+        self.grouper = grouper
+        self.sort = None if sort is None else functools.partial(sorted,
+                                                                key=key_sort,
+                                                                reverse=reverse)
+
+    @classmethod
+    def groupsof(cls, num=1, sort=None, reverse=False):
+        # copied from: https://docs.python.org/3/library/itertools.html#itertools.zip_longest
+        def grouper(jobs):
+            args = [iter(jobs)] * num
+            return zip_longest(*args)
+
+        return cls(grouper, sort, reverse)
+
+    @classmethod
+    def groupby(cls, key=None, default=None, sort=None, reverse=False):
+        if isinstance(key, six.string_types):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+
+        elif isinstance(key, Iterable):
+            if default is None:
+                def keyfunction(job):
+                    return tuple(job.sp[k] for k in key)
+            else:
+                def keyfunction(job):
+                    return tuple(job.sp.get(k, default) for k in key)
```

Comment:
If the user, uses a generator, we will consume it on the first call of the function. We need to convert to a list first to prevent that from happening.